R/dash.R

```diff
@@ -328,6 +301,7 @@ Dash <- R6::R6Class(
     },
     layout_set = function(...) {
       private$layout <- if (is.function(..1)) ..1 else list(...)
+      lay <- private$layout_render()
```

Comment:
This line confuses me a bit, although it might just be my misunderstanding of R. Where is `lay` being used? 
- Is this setting it to the class? 
- Does this also implicitly return `lay`?

Reply:
Also, why are we calling `layout_render` here? Could we leave a comment?

Reply:
Currently, if you don't assign `layout_render()` to an object, it returns the layout, which will display in the console.

There are a couple ways I could handle this -- I can wrap the return in `invisible`, so that it silently returns the layout, which can be assigned to a new object. Or, I can add a `silent` option which is `TRUE` by default and let the app developer decide.

Reply:
> Also, why are we calling `layout_render` here? Could we leave a comment?
I mentioned this last week, but I think it got lost in the shuffle. If we don't render the layout when it's set, I don't believe it will get rendered until a `GET` request unless `layout_get(render=TRUE)` is called.

Reply:
> it returns the layout, which will display in the console.

Because if you didn't return the object but still needed to call `private$layout_render()`, the code would look like:
```
layout_set = function(...) {
       private$layout <- if (is.function(..1)) ..1 else list(...)
       private$layout_render()
}
```

which would actually return `private$layout_render()` (since it's the last line of the function).

But if we didn't need to call `private$layout_render()`, then the function would just be 
```
layout_set = function(...) {
       private$layout <- if (is.function(..1)) ..1 else list(...)
}
```
which wouldn't return anything and be OK.

Am I understanding this correctly?

In Python, there is a convention to call "unused / do not care" arguments `_` (https://hackernoon.com/understanding-the-underscore-of-python-309d1a029edc). Not sure if you can assign variables to `_` in R or if R has that convention, but this codeblock might be easier for other folks to understand if it was written:
```
layout_set = function(...) {
       private$layout <- if (is.function(..1)) ..1 else list(...)
       # call layout_render() in order to  [...] but don't return its response
       _ <- private$layout_render()
}
```

Reply:
anyway, this isn't blocking, just trying to understand things a bit better!

Reply:
Oh, no worries -- I appreciate your questions because they often lead me to a better understanding of the way Dash works.

> This line confuses me a bit, although it might just be my misunderstanding of R. Where is `lay` being used?
> 
> * Is this setting it to the class?
> * Does this also implicitly return `lay`?

The code (which I've since modified) does not set class attributes or implicitly return `lay` since this line isn't the last one in the function. I've since wrapped `layout_render()` inside a call to `invisible`, which I think has the same effect as assignment to `_` in Python. (I actually did not know about that, but it's a cool feature of which I should be aware.)

I've added a brief comment to explain what's happening here also.

Reply:
Yep, would be nice to understand at some point why this call is necessary (for that matter, I don't quite get why we have a [nearly identical call](https://github.com/plotly/dash/blob/5b155f1e3690904a938040a84fdf2d4a20682e99/dash/dash.py#L297) on the Python side 😏) but thanks for making the intention clearer with `invisible`.

Reply:
fixed in f9c9174045abfea1593db06f0c5f4ce77a4a1b2c

Reply:
Missed a question or two here:

>which would actually return `private$layout_render()` (since it's the last line of the function).

Yes, that's right.

>But if we didn't need to call `private$layout_render()`, then the function would just be

Yep, got it -- this line basically says _if the first argument of `...` is a function, assign it to `layout`, otherwise wrap all the arguments into a list and assign that into `layout`.